src/Features/Marketing.php

```diff
@@ -0,0 +1,196 @@
+	 * @return array
+	 */
+	public function preload_options( $options ) {
+		$options[] = 'woocommerce_marketing_overview_welcome_hidden';
+
+		return $options;
+	}
+
+	/**
+	 * Add settings for marketing feature.
+	 *
+	 * @param array $settings Component settings.
+	 * @return array
+	 */
+	public function component_settings( $settings ) {
+		// Bail early if not on a wc-admin powered page.
+		if ( ! Loader::is_admin_page() ) {
+			return $settings;
+		}
+
+		$settings['marketing']['installedExtensions'] = InstalledExtensions::get_data();
+		$settings['marketing']['connectNonce']        = wp_create_nonce( 'connect' );
+
+		return $settings;
+	}
+
+	/**
+	 * Load recommended plugins from WooCommerce.com
+	 *
+	 * @return array
+	 */
+	public function get_recommended_plugins() {
+		$plugins = get_transient( self::RECOMMENDED_PLUGINS_TRANSIENT );
+
+		if ( false === $plugins ) {
+			// TODO update placeholder URL.
+			$request = wp_remote_get( 'https://ephemeral-findingsimple-20200320.atomicsites.blog/wp-json/wccom/marketing-tab/1.0/recommendations.json' );
+			$plugins = [];
+
+			if ( ! is_wp_error( $request ) && 200 === $request['response']['code'] ) {
+				$plugins = json_decode( $request['body'], true );
+			}
+
+			// Cache an empty result to avoid repeated failed requests.
+			set_transient( self::RECOMMENDED_PLUGINS_TRANSIENT, $plugins, DAY_IN_SECONDS );
+		}
+
+		return array_values( $plugins );
+	}
+
+	/**
+	 * Load knowledge base posts from WooCommerce.com
+	 *
+	 * @return array
+	 */
+	public function get_knowledge_base_posts() {
+		$posts = get_transient( self::KNOWLEDGE_BASE_TRANSIENT );
+
+		if ( false === $posts ) {
+			$request_url = add_query_arg(
+				array(
+					'categories' => 1744,
+					'page'       => 1,
+					'per_page'   => 8,
+					'_embed'     => 1,
+				),
+				'https://woocommerce.com/wp-json/wp/v2/posts'
+			);
+			$request = wp_remote_get( $request_url );
+			$posts   = [];
+
+			if ( ! is_wp_error( $request ) && 200 === $request['response']['code'] ) {
+				$raw_posts = json_decode( $request['body'], true );
+
+				foreach ( $raw_posts as $raw_post ) {
+					$post = [
+						'title'         => html_entity_decode( $raw_post['title']['rendered'] ),
+						'date'          => $raw_post['date_gmt'],
+						'link'          => $raw_post['link'],
+						'author_name'   => isset( $raw_post['author_name'] ) ? html_entity_decode( $raw_post['author_name'] ) : '',
+						'author_avatar' => isset( $raw_post['author_avatar_url'] ) ? $raw_post['author_avatar_url'] : '',
+					];
+
+					$featured_media = $raw_post['_embedded']['wp:featuredmedia'];
+
+					if ( count( $featured_media ) > 0 ) {
+						$image         = current( $featured_media );
+						$post['image'] = add_query_arg(
+							array(
+								'resize' => '650,340',
+								'crop'   => 1,
+							),
+							$image['source_url']
+						);
+					}
+
+					$posts[] = $post;
+				}
+			}
+
+			set_transient( self::KNOWLEDGE_BASE_TRANSIENT, $posts, DAY_IN_SECONDS );
```

Comment:
So one problem with this approach is if there is an error with the request (WooCommerce is temporarily down, or under heavy traffic) and posts aren't retrieved, then you'll be setting an empty array to the posts for another day.  I wonder if it'd be better here to:

- save posts to a regular wp option.
- use transient to decide whether to refresh the posts.
- if the request results in no posts, don't replace posts in the option, and set the transient for a shorter duration than a day (so a re-attempt can be made sooner).
- if the request results in posts (no problems with the request), then replace the posts cached in the option.

I think the above will prevent the (admittedly likely edgecase, but still) sites potentially ending up with no posts showing here.

Reply:
Good idea. I'll create an separate issue to follow this up / investigate.